s3-source-connector/src/main/java/io/aiven/kafka/connect/s3/source/utils/SourceRecordIterator.java

```diff
@@ -63,170 +54,119 @@ public final class SourceRecordIterator implements Iterator<S3SourceRecord> {
     public boolean hasNext() {
-        return recordIterator.hasNext() || objectListIterator.hasNext();
+        while (!outer.hasNext() && inner.hasNext()) {
+            outer = convert(inner.next()).iterator();
+        }
+        return outer.hasNext();
     }
 
     @Override
     public S3SourceRecord next() {
-        if (!recordIterator.hasNext()) {
-            nextS3Object();
-        }
-
-        if (!recordIterator.hasNext()) {
-            // If there are still no records, return null or throw an exception
-            return null; // Or throw new NoSuchElementException();
-        }
-
-        return recordIterator.next();
+        return outer.next();
     }
 
     @Override
     public void remove() {
         throw new UnsupportedOperationException("This iterator is unmodifiable");
     }
 
+    /**
+     * Converts the S3Object into stream of S3SourceRecords.
+     *
+     * @param s3Object
+     *            the S3Object to read data from.
+     * @return a stream of S3SourceRecords created from the input stream of the S3Object.
+     */
+    private Stream<S3SourceRecord> convert(final S3Object s3Object) {
+
+        final Map<String, Object> partitionMap = ConnectUtils.getPartitionMap(topic, partitionId, bucketName);
+        final long recordCount = offsetManager.recordsProcessedForObjectKey(partitionMap, s3Object.key());
+
+        // Optimizing without reading stream again.
+        if (transformer instanceof ByteArrayTransformer && recordCount > 0) {
+            return Stream.empty();
+        }
+
+        final SchemaAndValue keyData = transformer.getKeyData(s3Object.key(), topic, s3SourceConfig);
+
+        return transformer
+                .getRecords(sourceClient.getObject(s3Object.key()), topic, partitionId, s3SourceConfig, recordCount)
+                .map(new Mapper(partitionMap, recordCount, keyData, s3Object.key()));
+    }
+
+    /**
+     * maps the data from the @{link Transformer} stream to an S3SourceRecord given all the additional data required.
+     */
+    class Mapper implements Function<SchemaAndValue, S3SourceRecord> {
+        /**
+         * The partition map
+         */
+        private final Map<String, Object> partitionMap;
+        /**
+         * The record number for the record being created.
+         */
+        private long recordCount;
+        /**
+         * The schema and value for the key
+         */
+        private final SchemaAndValue keyData;
+        /**
+         * The object key from S3
+         */
+        private final String objectKey;
+
+        public Mapper(final Map<String, Object> partitionMap, final long recordCount, final SchemaAndValue keyData,
+                final String objectKey) {
+            this.partitionMap = partitionMap;
+            this.recordCount = recordCount;
+            this.keyData = keyData;
+            this.objectKey = objectKey;
+        }
+
+        @Override
+        public S3SourceRecord apply(final SchemaAndValue value) {
```

Comment:
```suggestion
        public S3SourceRecord apply(final SchemaAndValue valueData) {
```
to keep it aligned with keyData